python/src/opendp/_extrinsics/microdata/faithfulness.py

```diff
@@ -0,0 +1,127 @@
+    :param similarity: function measuing the similarity between two records
+            or a dictionary of weights for each column scaling the absolute difference
+
+    :returns a Transformation that computes a tuple of (mean, S, Vt)
+    """
+
+    np = import_optional_dependency("numpy")
+    scipy = import_optional_dependency("scipy")
+    sklearn_neighbors = import_optional_dependency("sklearn.neighbors")
+    igraph = import_optional_dependency("igraph")
+
+    dp.assert_features("contrib", "floating-point")
+
+    if "LazyFrameDomain" not in str(input_domain.type):
+        raise ValueError("input_domain must be a LazyFrame domain")
+
+    # TODO: how to enfornce public knowlege of dataset size, if at all?
+    # if (input_domain.descriptor.get("size")) is None:
+    #     raise ValueError("input_domain's size must be known")
+
+    # assert input_domain.member(reference_dataset.lazy())
+
+    if input_metric != dp.symmetric_distance():
+        raise ValueError("input metric must be symmetric distance")
+
+    if callable(similarity):
+        similarity_metric = similarity
+
+    elif isinstance(similarity, dict):
+        assert reference_dataset.columns == list(similarity.keys())
+        metric_weights = np.array(list(similarity.values()))
+
+        def similarity_metric(xi, yj):
+            return np.sum(np.abs((xi - yj)) * metric_weights)
+
+    else:
+        raise ValueError("similarity must be a callable or a dictionary")
+
+    def similarity_fn(X, Y):
+        # TODO: consider using the underline scipy implementation directly
+        neigh = sklearn_neighbors.NearestNeighbors(metric=similarity_metric, p=1, radius=1)
+        neigh.fit(X)
+        similar_record_indices = neigh.radius_neighbors(Y, return_distance=False)
+        return similar_record_indices
+
+    def _compute_optimal_faithfulness_cardinality(dataset):        
+        num_records = len(reference_dataset)
+
+        similar_record_indices = similarity_fn(dataset.collect(), reference_dataset)
+
+        # build bipartite graph and find maximum matching
+
+        gmat = scipy.sparse.lil_matrix((num_records, num_records))
+
+        for row, cols in enumerate(similar_record_indices):
+            if len(cols):
+                gmat[row, sorted(cols)] = np.ones(len(cols), dtype=int)
+
+        gmat = gmat.tocsr()
+
+        nz = gmat.nonzero()
+
+        edges = zip(nz[0], nz[1] + num_records)
+
+        bipartite_mask = [False] * num_records + [True] * num_records
+        g = igraph.Graph.Bipartite(bipartite_mask, edges, directed=False)
+
+        matched_vertices = np.array(g.maximum_bipartite_matching().matching)[:num_records]
+
+        second_matched_mask = matched_vertices[:num_records] != -1
+        second_matched = np.nonzero(second_matched_mask)[0]
+        first_matched = matched_vertices[second_matched] - num_records
+
+        assert len(first_matched) == len(second_matched)
```

Comment:
If this is an algorithm check, and it would just surface a bug in our code if it failed, then this is ok. Still might add a second argument with a message to help the person who gets the bug report.